### PR TITLE
Mpuncel/daemon set status

### PR DESCRIPTION
**Pull request by mpuncel**

### CODE REVIEW

**Comment by mpuncel** [comment]
this contains commits from #914 so review that first

File: pkg/ds/farm.go [hunk 1]
```diff
@@ -614,29 +625,37 @@ func (dsf *Farm) spawnDaemonSet(
...
 			// The session probably either expired or there was probably a network
 			// error, so the rest will probably fail
 			dsf.handleSessionExpiry(dsFields, dsLogger, err)
-			dsf.releaseLock(dsUnlocker)
+			return false
+		}
+		if !ok {
+			// Lock was either already acquired by another farm or it was Acquired
+			// by this farm
+			dsf.logger.Infof("Lock on daemon set '%v' was already acquired by another ds farm: %s", dsFields.ID, transaction.TxnErrorsToString(resp.Errors))
```
Comment (rudle) [inline pkg/ds/farm.go]: `Errorf()`

File: pkg/ds/daemon_set.go [hunk 2]
```diff
@@ -224,12 +225,24 @@ func (ds *daemonSet) GetNodeSelector() klabels.Selector {
 	return ds.DaemonSet.NodeSelector
 }
 
+func (ds *daemonSet) Manifest() manifest.Manifest {
+	ds.mu.Lock()
+	defer ds.mu.Unlock()
```
Comment (rudle) [inline pkg/ds/daemon_set.go]: Why read locks? Is this same mutex used to guard writes too?

File: pkg/ds/daemon_set.go [hunk 3]
```diff
@@ -34,7 +37,8 @@ const (
 )
 
 var (
-	DefaultRetryInterval = 5 * time.Minute
+	DefaultRetryInterval  = 5 * time.Minute
+	statusWritingInterval = 15 * time.Second
```
Comment (rudle) [inline pkg/ds/daemon_set.go]: Make this a config flag?

File: pkg/ds/farm.go (hunk 1 above)
Comment (mpuncel) [reply pkg/ds/farm.go]: I don't think it's really an error, it just means that the daemon set is already claimed by another farm. we expect that to happen often

File: pkg/ds/daemon_set.go (hunk 3 above)
Comment (mpuncel) [reply pkg/ds/daemon_set.go]: 👍

File: pkg/ds/daemon_set.go (hunk 2 above)
Comment (mpuncel) [reply pkg/ds/daemon_set.go]: yeah, https://github.com/square/p2/blob/master/pkg/ds/daemon_set.go#L313

Comment (mpuncel) [reply pkg/ds/daemon_set.go]: although maybe there's only one goroutine acting here?

**Comment by mpuncel** [comment]
this has undergone a bit of turmoil since the last review, so not sure if i should still consider it approved.

the main change is that we don't CAS the status values anymore, that proved challenging because the `queryMeta` gives you the last GLOBAL index of consul not the modify index of the key you fetched. I didn't want to tear up our APIs to get the modify index out of the `api.KVPair`. Instead i'm just using a "set" not a "cas" now which I believe to be no less thorough because we do check we have the daemon set lock still and that combined with single goroutine writes should give us the same guarantees

**Comment by rudle** [comment]
I agree with your reasoning. 

I recall that CAS was in the API before transactions were released and it remains around as a less expensive approach to synchronization.

**Comment by rudle** [comment]
:shipit:

**Comment by mpuncel** [comment]
Added a commit to this because I noticed in testing that `replication_in_progress` never became false in the status. Added an `InProgress()` function to replications to handle that

**Comment by mpuncel** [comment]
reviewed the final commit with @rudle in person

**Pull request merged**